File: mobile/ios/Features/JobDiscovery/JobModels.swift
import Foundation

enum JobWorkMode: String, TolerantEnum, CaseIterable {
    case remote, hybrid, onSite = "on_site", flexible, unknown
    var label: String { switch self { case .remote: "远程"; case .hybrid: "混合"; case .onSite: "现场"; case .flexible: "灵活"; case .unknown: "未说明" } }
}

enum JobEmploymentType: String, TolerantEnum, CaseIterable {
    case fullTime = "full_time", partTime = "part_time", contract, internship, freelance, temporary, unknown
    var label: String { switch self { case .fullTime: "全职"; case .partTime: "兼职"; case .contract: "合同"; case .internship: "实习"; case .freelance: "自由职业"; case .temporary: "临时"; case .unknown: "未说明" } }
}

enum JobSeniority: String, TolerantEnum, CaseIterable {
    case intern, junior, mid, senior, lead, manager, director, executive, unknown
    var label: String { switch self { case .intern: "实习"; case .junior: "初级"; case .mid: "中级"; case .senior: "高级"; case .lead: "负责人"; case .manager: "经理"; case .director: "总监"; case .executive: "高管"; case .unknown: "未说明" } }
}

enum SalaryPeriod: String, TolerantEnum, CaseIterable {
    case hourly, daily, monthly, annual, project, unknown
}

enum JobEligibility: String, TolerantEnum {
    case eligible, notEligible = "not_eligible", unknown
}

enum JobFeedbackType: String, TolerantEnum, CaseIterable {
    case relevant, notRelevant = "not_relevant", notAJob = "not_a_job", duplicate, expired, scam, wrongExtraction = "wrong_extraction", unknown
    var label: String { switch self { case .relevant: "适合我"; case .notRelevant: "不适合"; case .notAJob: "不是招聘"; case .duplicate: "重复"; case .expired: "已过期"; case .scam: "疑似诈骗"; case .wrongExtraction: "提取错误"; case .unknown: "其他" } }
}

struct JobMatch: Codable, Sendable, Hashable {
    var eligibility: JobEligibility
    var matchScore: Int
    var matchedReasons: [String]
    var mismatchReasons: [String]
    var unknownConstraints: [String]
    var scoreBreakdown: [String: Int]
}

struct JobSource: Codable, Sendable, Identifiable, Hashable {
    var id: UUID
    var channel: IMChannel
    var chatName: String?
    var authorName: String?
    var postedAt: Date
    var sourceMessageUrl: URL?
    var reliabilityScore: Double
}

struct JobOpportunity: Codable, Sendable, Identifiable, Hashable {
    var opportunityId: UUID
    var jobTitle: String
    var companyName: String?
    var sourceChannel: IMChannel
    var sourceChatName: String?
    var postedAt: Date
    var locationText: String?
    var countryCode: String?
    var city: String?
    var workMode: JobWorkMode
    var employmentType: JobEmploymentType
    var seniority: JobSeniority
    var salaryRaw: String?
    var salaryMin: Double?
    var salaryMax: Double?
    var salaryCurrency: String?
    var salaryPeriod: SalaryPeriod
    var requiredSkills: [String]
    var degreeLevel: String?
    var englishLevel: String?
    var visaSponsorship: Bool?
    var applicationDeadline: Date?
    var sourceReliabilityScore: Double
    var extractionConfidence: Double
    var sourceCount: Int
    var conflictingSourceData: Bool
    var complianceFlags: [String]
    var isExpired: Bool
    var match: JobMatch?
    var id: UUID { opportunityId }
}

struct JobContactMethod: Codable, Sendable, Hashable { var type: String; var value: String }

struct JobOpportunityDetail: Codable, Sendable, Identifiable, Hashable {
    var opportunityId: UUID
    var jobTitle: String
    var companyName: String?
    var sourceChannel: IMChannel
    var sourceChatName: String?
    var postedAt: Date
    var locationText: String?
    var countryCode: String?
    var city: String?
    var workMode: JobWorkMode
    var employmentType: JobEmploymentType
    var seniority: JobSeniority
    var salaryRaw: String?
    var salaryMin: Double?
    var salaryMax: Double?
    var salaryCurrency: String?
    var salaryPeriod: SalaryPeriod
    var requiredSkills: [String]
    var degreeLevel: String?
    var englishLevel: String?
    var visaSponsorship: Bool?
    var applicationDeadline: Date?
    var sourceReliabilityScore: Double
    var extractionConfidence: Double
    var sourceCount: Int
    var conflictingSourceData: Bool
    var complianceFlags: [String]
    var isExpired: Bool
    var match: JobMatch?
    var sourceMessageUrl: URL?
    var sourceAuthorName: String?
    var department: String?
    var companyIndustry: String?
    var companyStage: String?
    var timezone: String?
    var salaryNegotiable: Bool?
    var equityMentioned: Bool?
    var requirementsSummary: String?
    var preferredSkills: [String]
    var minimumYearsExperience: Double?
    var maximumYearsExperience: Double?
    var degreeRequired: Bool?
    var degreeField: String?
    var otherLanguageRequirements: [String]
    var workAuthorizationText: String?
    var relocationSupport: Bool?
    var ageRequirementText: String?
    var ageRequirementPresent: Bool
    var applicationUrl: URL?
    var contactMethods: [JobContactMethod]
    var missingFields: [String]
    var fieldEvidence: [String: String]
    var rawExcerpt: String
    var expiredReason: String?
    var sources: [JobSource]
    var id: UUID { opportunityId }
}

struct JobSearchProfile: Codable, Sendable, Identifiable, Hashable {
    var id: UUID
    var name: String
    var isDefault: Bool
    var enabled: Bool
    var targetRoles: [String]
    var excludedRoles: [String]
    var targetIndustries: [String]
    var preferredSeniority: [JobSeniority]
    var candidateSkills: [String]
    var yearsExperience: Double?
    var educationLevel: String?
    var englishLevel: String?
    var otherLanguages: [String]
    var preferredCountries: [String]
    var preferredCities: [String]
    var preferredTimezones: [String]
    var workModes: [JobWorkMode]
    var employmentTypes: [JobEmploymentType]
    var minimumSalary: Double?
    var salaryCurrency: String?
    var salaryPeriod: SalaryPeriod?
    var visaSponsorshipRequired: Bool?
    var relocationAcceptable: Bool?
    var requiredKeywords: [String]
    var preferredKeywords: [String]
    var excludedKeywords: [String]
    var requireSalaryDisclosed: Bool
    var minimumMatchScore: Int
    var notificationEnabled: Bool
    var createdAt: Date
    var updatedAt: Date
}

struct JobSearchProfileWrite: Codable, Sendable, Hashable {
    var name = ""
    var isDefault = false
    var enabled = true
    var targetRoles: [String] = []
    var excludedRoles: [String] = []
    var targetIndustries: [String] = []
    var preferredSeniority: [JobSeniority] = []
    var candidateSkills: [String] = []
    var yearsExperience: Double?
    var educationLevel: String?
    var englishLevel: String?
    var otherLanguages: [String] = []
    var preferredCountries: [String] = []
    var preferredCities: [String] = []
    var preferredTimezones: [String] = []
    var workModes: [JobWorkMode] = []
    var employmentTypes: [JobEmploymentType] = []
    var minimumSalary: Double?
    var salaryCurrency: String?
    var salaryPeriod: SalaryPeriod?
    var visaSponsorshipRequired: Bool?
    var relocationAcceptable: Bool?
    var requiredKeywords: [String] = []
    var preferredKeywords: [String] = []
    var excludedKeywords: [String] = []
    var requireSalaryDisclosed = false
    var minimumMatchScore = 0
    var notificationEnabled = false

    init() {}
    init(_ profile: JobSearchProfile) {
        name = profile.name; isDefault = profile.isDefault; enabled = profile.enabled
        targetRoles = profile.targetRoles; excludedRoles = profile.excludedRoles
        targetIndustries = profile.targetIndustries; preferredSeniority = profile.preferredSeniority
        candidateSkills = profile.candidateSkills; yearsExperience = profile.yearsExperience
        educationLevel = profile.educationLevel; englishLevel = profile.englishLevel
        otherLanguages = profile.otherLanguages; preferredCountries = profile.preferredCountries
        preferredCities = profile.preferredCities; preferredTimezones = profile.preferredTimezones
        workModes = profile.workModes; employmentTypes = profile.employmentTypes
        minimumSalary = profile.minimumSalary; salaryCurrency = profile.salaryCurrency
        salaryPeriod = profile.salaryPeriod; visaSponsorshipRequired = profile.visaSponsorshipRequired
        relocationAcceptable = profile.relocationAcceptable; requiredKeywords = profile.requiredKeywords
        preferredKeywords = profile.preferredKeywords; excludedKeywords = profile.excludedKeywords
        requireSalaryDisclosed = profile.requireSalaryDisclosed
        minimumMatchScore = profile.minimumMatchScore; notificationEnabled = profile.notificationEnabled
    }
}

struct JobSearchProfilePreview: Codable, Sendable {
    var name: String
    var isDefault: Bool
    var enabled: Bool
    var targetRoles: [String]
    var excludedRoles: [String]
    var targetIndustries: [String]
    var preferredSeniority: [JobSeniority]
    var candidateSkills: [String]
    var yearsExperience: Double?
    var educationLevel: String?
    var englishLevel: String?
    var otherLanguages: [String]
    var preferredCountries: [String]
    var preferredCities: [String]
    var preferredTimezones: [String]
    var workModes: [JobWorkMode]
    var employmentTypes: [JobEmploymentType]
    var minimumSalary: Double?
    var salaryCurrency: String?
    var salaryPeriod: SalaryPeriod?
    var visaSponsorshipRequired: Bool?
    var relocationAcceptable: Bool?
    var requiredKeywords: [String]
    var preferredKeywords: [String]
    var excludedKeywords: [String]
    var requireSalaryDisclosed: Bool
    var minimumMatchScore: Int
    var notificationEnabled: Bool
    var requiresConfirmation: Bool

    var write: JobSearchProfileWrite {
        var value = JobSearchProfileWrite()
        value.name = name; value.isDefault = isDefault; value.enabled = enabled
        value.targetRoles = targetRoles; value.excludedRoles = excludedRoles
        value.targetIndustries = targetIndustries; value.preferredSeniority = preferredSeniority
        value.candidateSkills = candidateSkills; value.yearsExperience = yearsExperience
        value.educationLevel = educationLevel; value.englishLevel = englishLevel
        value.otherLanguages = otherLanguages; value.preferredCountries = preferredCountries
        value.preferredCities = preferredCities; value.preferredTimezones = preferredTimezones
        value.workModes = workModes; value.employmentTypes = employmentTypes
        value.minimumSalary = minimumSalary; value.salaryCurrency = salaryCurrency
        value.salaryPeriod = salaryPeriod; value.visaSponsorshipRequired = visaSponsorshipRequired
        value.relocationAcceptable = relocationAcceptable; value.requiredKeywords = requiredKeywords
        value.preferredKeywords = preferredKeywords; value.excludedKeywords = excludedKeywords
        value.requireSalaryDisclosed = requireSalaryDisclosed
        value.minimumMatchScore = minimumMatchScore; value.notificationEnabled = notificationEnabled
        return value
    }
}

struct JobsPage: Codable, Sendable {
    var items: [JobOpportunity]
    var total: Int
    var limit: Int
    var offset: Int
    var filterSummary: [String: JSONValue]
    var profile: JobSearchProfile?
}

struct JobProfileParseRequest: Encodable, Sendable { var text: String }
struct JobFeedbackRequest: Encodable, Sendable { var feedbackType: JobFeedbackType; var note: String? }
struct JobFeedbackResponse: Decodable, Sendable { var id: UUID; var feedbackType: JobFeedbackType; var note: String?; var updatedAt: Date }

File: mobile/android/app/src/main/kotlin/com/codeiy/im/model/JobModels.kt
package com.codeiy.im.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
enum class JobWorkMode(val label: String) {
    @SerialName("remote") REMOTE("远程"),
    @SerialName("hybrid") HYBRID("混合"),
    @SerialName("on_site") ON_SITE("现场"),
    @SerialName("flexible") FLEXIBLE("灵活"),
    UNKNOWN("未说明"),
}

@Serializable
enum class JobEmploymentType(val label: String) {
    @SerialName("full_time") FULL_TIME("全职"),
    @SerialName("part_time") PART_TIME("兼职"),
    @SerialName("contract") CONTRACT("合同"),
    @SerialName("internship") INTERNSHIP("实习"),
    @SerialName("freelance") FREELANCE("自由职业"),
    @SerialName("temporary") TEMPORARY("临时"),
    UNKNOWN("未说明"),
}

@Serializable
enum class JobSeniority(val label: String) {
    @SerialName("intern") INTERN("实习"),
    @SerialName("junior") JUNIOR("初级"),
    @SerialName("mid") MID("中级"),
    @SerialName("senior") SENIOR("高级"),
    @SerialName("lead") LEAD("负责人"),
    @SerialName("manager") MANAGER("经理"),
    @SerialName("director") DIRECTOR("总监"),
    @SerialName("executive") EXECUTIVE("高管"),
    UNKNOWN("未说明"),
}

@Serializable
enum class SalaryPeriod {
    @SerialName("hourly") HOURLY,
    @SerialName("daily") DAILY,
    @SerialName("monthly") MONTHLY,
    @SerialName("annual") ANNUAL,
    @SerialName("project") PROJECT,
    UNKNOWN,
}

@Serializable
enum class JobEligibility {
    @SerialName("eligible") ELIGIBLE,
    @SerialName("not_eligible") NOT_ELIGIBLE,
    UNKNOWN,
}

@Serializable
enum class JobFeedbackType(val label: String) {
    @SerialName("relevant") RELEVANT("适合我"),
    @SerialName("not_relevant") NOT_RELEVANT("不适合"),
    @SerialName("not_a_job") NOT_A_JOB("不是招聘"),
    @SerialName("duplicate") DUPLICATE("重复"),
    @SerialName("expired") EXPIRED("已过期"),
    @SerialName("scam") SCAM("疑似诈骗"),
    @SerialName("wrong_extraction") WRONG_EXTRACTION("提取错误"),
    UNKNOWN("其他"),
}

@Serializable
data class JobMatch(
    val eligibility: JobEligibility = JobEligibility.UNKNOWN,
    val matchScore: Int = 0,
    val matchedReasons: List<String> = emptyList(),
    val mismatchReasons: List<String> = emptyList(),
    val unknownConstraints: List<String> = emptyList(),
    val scoreBreakdown: Map<String, Int> = emptyMap(),
)

@Serializable
data class JobSource(
    val id: String,
    val channel: IMChannel = IMChannel.UNKNOWN,
    val chatName: String? = null,
    val authorName: String? = null,
    val postedAt: String = "",
    val sourceMessageUrl: String? = null,
    val reliabilityScore: Double = 0.0,
)

@Serializable
data class JobOpportunity(
    val opportunityId: String,
    val jobTitle: String = "",
    val companyName: String? = null,
    val sourceChannel: IMChannel = IMChannel.UNKNOWN,
    val sourceChatName: String? = null,
    val postedAt: String = "",
    val locationText: String? = null,
    val countryCode: String? = null,
    val city: String? = null,
    val workMode: JobWorkMode = JobWorkMode.UNKNOWN,
    val employmentType: JobEmploymentType = JobEmploymentType.UNKNOWN,
    val seniority: JobSeniority = JobSeniority.UNKNOWN,
    val salaryRaw: String? = null,
    val salaryMin: Double? = null,
    val salaryMax: Double? = null,
    val salaryCurrency: String? = null,
    val salaryPeriod: SalaryPeriod = SalaryPeriod.UNKNOWN,
    val requiredSkills: List<String> = emptyList(),
    val degreeLevel: String? = null,
    val englishLevel: String? = null,
    val visaSponsorship: Boolean? = null,
    val applicationDeadline: String? = null,
    val sourceReliabilityScore: Double = 0.0,
    val extractionConfidence: Double = 0.0,
    val sourceCount: Int = 1,
    val conflictingSourceData: Boolean = false,
    val complianceFlags: List<String> = emptyList(),
    val isExpired: Boolean = false,
    val match: JobMatch? = null,
)

@Serializable
data class JobOpportunityDetail(
    val opportunityId: String,
    val jobTitle: String = "",
    val companyName: String? = null,
    val sourceChannel: IMChannel = IMChannel.UNKNOWN,
    val sourceChatName: String? = null,
    val postedAt: String = "",
    val locationText: String? = null,
    val countryCode: String? = null,
    val city: String? = null,
    val workMode: JobWorkMode = JobWorkMode.UNKNOWN,
    val employmentType: JobEmploymentType = JobEmploymentType.UNKNOWN,
    val seniority: JobSeniority = JobSeniority.UNKNOWN,
    val salaryRaw: String? = null,
    val salaryMin: Double? = null,
    val salaryMax: Double? = null,
    val salaryCurrency: String? = null,
    val salaryPeriod: SalaryPeriod = SalaryPeriod.UNKNOWN,
    val requiredSkills: List<String> = emptyList(),
    val degreeLevel: String? = null,
    val englishLevel: String? = null,
    val visaSponsorship: Boolean? = null,
    val applicationDeadline: String? = null,
    val sourceReliabilityScore: Double = 0.0,
    val extractionConfidence: Double = 0.0,
    val sourceCount: Int = 1,
    val conflictingSourceData: Boolean = false,
    val complianceFlags: List<String> = emptyList(),
    val isExpired: Boolean = false,
    val match: JobMatch? = null,
    val sourceMessageUrl: String? = null,
    val sourceAuthorName: String? = null,
    val department: String? = null,
    val companyIndustry: String? = null,
    val companyStage: String? = null,
    val timezone: String? = null,
    val salaryNegotiable: Boolean? = null,
    val equityMentioned: Boolean? = null,
    val requirementsSummary: String? = null,
    val preferredSkills: List<String> = emptyList(),
    val minimumYearsExperience: Double? = null,
    val maximumYearsExperience: Double? = null,
    val degreeRequired: Boolean? = null,
    val degreeField: String? = null,
    val otherLanguageRequirements: List<String> = emptyList(),
    val workAuthorizationText: String? = null,
    val relocationSupport: Boolean? = null,
    val ageRequirementText: String? = null,
    val ageRequirementPresent: Boolean = false,
    val applicationUrl: String? = null,
    val contactMethods: List<Map<String, String>> = emptyList(),
    val missingFields: List<String> = emptyList(),
    val fieldEvidence: Map<String, String> = emptyMap(),
    val rawExcerpt: String = "",
    val expiredReason: String? = null,
    val sources: List<JobSource> = emptyList(),
)

@Serializable
data class JobSearchProfileWrite(
    val name: String = "",
    val isDefault: Boolean = false,
    val enabled: Boolean = true,
    val targetRoles: List<String> = emptyList(),
    val excludedRoles: List<String> = emptyList(),
    val targetIndustries: List<String> = emptyList(),
    val preferredSeniority: List<JobSeniority> = emptyList(),
    val candidateSkills: List<String> = emptyList(),
    val yearsExperience: Double? = null,
    val educationLevel: String? = null,
    val englishLevel: String? = null,
    val otherLanguages: List<String> = emptyList(),
    val preferredCountries: List<String> = emptyList(),
    val preferredCities: List<String> = emptyList(),
    val preferredTimezones: List<String> = emptyList(),
    val workModes: List<JobWorkMode> = emptyList(),
    val employmentTypes: List<JobEmploymentType> = emptyList(),
    val minimumSalary: Double? = null,
    val salaryCurrency: String? = null,
    val salaryPeriod: SalaryPeriod? = null,
    val visaSponsorshipRequired: Boolean? = null,
    val relocationAcceptable: Boolean? = null,
    val requiredKeywords: List<String> = emptyList(),
    val preferredKeywords: List<String> = emptyList(),
    val excludedKeywords: List<String> = emptyList(),
    val requireSalaryDisclosed: Boolean = false,
    val minimumMatchScore: Int = 0,
    val notificationEnabled: Boolean = false,
)

@Serializable
data class JobSearchProfile(
    val id: String,
    val name: String = "",
    val isDefault: Boolean = false,
    val enabled: Boolean = true,
    val targetRoles: List<String> = emptyList(),
    val excludedRoles: List<String> = emptyList(),
    val targetIndustries: List<String> = emptyList(),
    val preferredSeniority: List<JobSeniority> = emptyList(),
    val candidateSkills: List<String> = emptyList(),
    val yearsExperience: Double? = null,
    val educationLevel: String? = null,
    val englishLevel: String? = null,
    val otherLanguages: List<String> = emptyList(),
    val preferredCountries: List<String> = emptyList(),
    val preferredCities: List<String> = emptyList(),
    val preferredTimezones: List<String> = emptyList(),
    val workModes: List<JobWorkMode> = emptyList(),
    val employmentTypes: List<JobEmploymentType> = emptyList(),
    val minimumSalary: Double? = null,
    val salaryCurrency: String? = null,
    val salaryPeriod: SalaryPeriod? = null,
    val visaSponsorshipRequired: Boolean? = null,
    val relocationAcceptable: Boolean? = null,
    val requiredKeywords: List<String> = emptyList(),
    val preferredKeywords: List<String> = emptyList(),
    val excludedKeywords: List<String> = emptyList(),
    val requireSalaryDisclosed: Boolean = false,
    val minimumMatchScore: Int = 0,
    val notificationEnabled: Boolean = false,
    val createdAt: String = "",
    val updatedAt: String = "",
) {
    fun toWrite() = JobSearchProfileWrite(
        name, isDefault, enabled, targetRoles, excludedRoles, targetIndustries,
        preferredSeniority, candidateSkills, yearsExperience, educationLevel, englishLevel,
        otherLanguages, preferredCountries, preferredCities, preferredTimezones, workModes,
        employmentTypes, minimumSalary, salaryCurrency, salaryPeriod, visaSponsorshipRequired,
        relocationAcceptable, requiredKeywords, preferredKeywords, excludedKeywords,
        requireSalaryDisclosed, minimumMatchScore, notificationEnabled,
    )
}

@Serializable
data class JobSearchProfilePreview(
    val name: String = "",
    val isDefault: Boolean = false,
    val enabled: Boolean = true,
    val targetRoles: List<String> = emptyList(),
    val excludedRoles: List<String> = emptyList(),
    val targetIndustries: List<String> = emptyList(),
    val preferredSeniority: List<JobSeniority> = emptyList(),
    val candidateSkills: List<String> = emptyList(),
    val yearsExperience: Double? = null,
    val educationLevel: String? = null,
    val englishLevel: String? = null,
    val otherLanguages: List<String> = emptyList(),
    val preferredCountries: List<String> = emptyList(),
    val preferredCities: List<String> = emptyList(),
    val preferredTimezones: List<String> = emptyList(),
    val workModes: List<JobWorkMode> = emptyList(),
    val employmentTypes: List<JobEmploymentType> = emptyList(),
    val minimumSalary: Double? = null,
    val salaryCurrency: String? = null,
    val salaryPeriod: SalaryPeriod? = null,
    val visaSponsorshipRequired: Boolean? = null,
    val relocationAcceptable: Boolean? = null,
    val requiredKeywords: List<String> = emptyList(),
    val preferredKeywords: List<String> = emptyList(),
    val excludedKeywords: List<String> = emptyList(),
    val requireSalaryDisclosed: Boolean = false,
    val minimumMatchScore: Int = 0,
    val notificationEnabled: Boolean = false,
    val requiresConfirmation: Boolean = true,
) {
    fun toWrite() = JobSearchProfileWrite(
        name, isDefault, enabled, targetRoles, excludedRoles, targetIndustries,
        preferredSeniority, candidateSkills, yearsExperience, educationLevel, englishLevel,
        otherLanguages, preferredCountries, preferredCities, preferredTimezones, workModes,
        employmentTypes, minimumSalary, salaryCurrency, salaryPeriod, visaSponsorshipRequired,
        relocationAcceptable, requiredKeywords, preferredKeywords, excludedKeywords,
        requireSalaryDisclosed, minimumMatchScore, notificationEnabled,
    )
}

@Serializable
data class JobsPage(
    val items: List<JobOpportunity> = emptyList(),
    val total: Int = 0,
    val limit: Int = 20,
    val offset: Int = 0,
    val filterSummary: Map<String, JsonElement> = emptyMap(),
    val profile: JobSearchProfile? = null,
)

@Serializable
data class JobProfileParseRequest(val text: String)

@Serializable
data class JobFeedbackRequest(val feedbackType: JobFeedbackType, val note: String? = null)

@Serializable
data class JobFeedbackResponse(
    val id: String,
    val feedbackType: JobFeedbackType,
    val note: String? = null,
    val updatedAt: String = "",
)

